lib/ansible/module_utils/hwc_utils.py

```diff
@@ -0,0 +1,319 @@
+                    fallback=(env_fallback, ['PASSWORD']),
+                ),
+                domain_name=dict(
+                    required=True, type='str',
+                    fallback=(env_fallback, ['DOMAIN_NAME']),
+                ),
+                project_name=dict(
+                    required=True, type='str',
+                    fallback=(env_fallback, ['PROJECT_NAME']),
+                ),
+                region=dict(
+                    required=True, type='str',
+                    fallback=(env_fallback, ['REGION']),
+                ),
+                timeouts=dict(type='dict', options=dict(
+                    create=dict(default='10m', type='str'),
+                    update=dict(default='10m', type='str'),
+                    delete=dict(default='10m', type='str'),
+                ), default={}),
+                id=dict(type='str')
+            )
+        )
+
+        super(HwcModule, self).__init__(*args, **kwargs)
+
+
+class DictComparison(object):
+    ''' This class takes in two dictionaries `a` and `b`.
+        These are dictionaries of arbitrary depth, but made up of standard
+        Python types only.
+        This differ will compare all values in `a` to those in `b`.
+        Note: Only keys in `a` will be compared. Extra keys in `b` will be ignored.
+        Note: On all lists, order does matter.
+    '''
+
+    def __init__(self, request):
+        self.request = request
+
+    def __eq__(self, other):
+        return self._compare_dicts(self.request, other.request)
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def _compare_dicts(self, dict1, dict2):
+        if len(dict1.keys()) != len(dict2.keys()):
+            return False
+
+        return all([
+            self._compare_value(dict1.get(k), dict2.get(k)) for k in dict1
+        ])
+
+    def _compare_lists(self, list1, list2):
+        """Takes in two lists and compares them."""
+        if len(list1) != len(list2):
+            return False
+
+        difference = []
+        for index in range(len(list1)):
+            value1 = list1[index]
+            if index < len(list2):
+                value2 = list2[index]
+                difference.append(self._compare_value(value1, value2))
+
+        return all(difference)
+
+    def _compare_value(self, value1, value2):
+        """
+        return: True: value1 is same as value2, otherwise False.
+        """
+        if not (value1 and value2):
+            return (not value1) and (not value2)
+
+        # Can assume non-None types at this point.
+        try:
+            if isinstance(value1, list):
+                return self._compare_lists(value1, value2)
+            elif isinstance(value1, dict):
+                return self._compare_dicts(value1, value2)
+            # Always use to_text values to avoid unicode issues.
+            else:
+                return to_text(value1) == to_text(value2)
```

Comment:
If you use `to_text(xxx, errors='surrogate_or_strict')` it won't throw exceptions.